ext/go-plugin/vagrant/plugin/io.go

```diff
@@ -0,0 +1,100 @@
+package plugin
+
+import (
+	"context"
+
+	"google.golang.org/grpc"
+
+	go_plugin "github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/vagrant/ext/go-plugin/vagrant"
+	"github.com/hashicorp/vagrant/ext/go-plugin/vagrant/plugin/proto"
+
+	"github.com/LK4D4/joincontext"
+)
+
+type IO interface {
+	vagrant.StreamIO
+}
+
+type IOPlugin struct {
+	go_plugin.NetRPCUnsupportedPlugin
+	Impl vagrant.StreamIO
+}
+
+type GRPCIOServer struct {
+	Impl vagrant.StreamIO
+}
+
+func (s *GRPCIOServer) Read(ctx context.Context, req *vagrant_proto.Identifier) (r *vagrant_proto.Content, err error) {
+	r = &vagrant_proto.Content{}
+	n := make(chan struct{})
+	go func() {
+		r.Value, err = s.Impl.Read(req.Name)
+		n <- struct{}{}
+	}()
+	select {
+	case <-ctx.Done():
+	case <-n:
```

Comment:
Nice rewrite on the context support !! !

Just FYI: another nice way of doing this could be to use [errgroup.WithContext](https://godoc.org/golang.org/x/sync/errgroup#WithContext); this way the Read error could be bubbled down too.

Also, may be it's more complicated than that; but we wouldn't have to do `go` + `select` if the `Read` was context aware to be cancelled.